#Домашнє завдання до теми "Comprehensions в Python"

Для наступних завдань у нас є наступний список:

In [59]:
numbers = [42, 51, 29, 2, 13, 44, 80, 66, 16]

1. Використовуючи цикл `for` cтворіть новий список `numbers_squared`, де кожен елемент списку `numbers` є домноженим на 2.

Нагадаю, що фактично коли ми з допомогою циклу отримуємо новий список, то ми спочатку створюємо пустий список, а потім додаємо в нього новий елемент на кожній ітерації.

Новий елемент буде в даному випадку результатом домножання чергового елемента вихідного списку `numbers` на 2.

In [60]:
numbers_squared = []
for n in numbers:
  numbers_squared.append(n*2)

In [61]:
print(numbers_squared)

[84, 102, 58, 4, 26, 88, 160, 132, 32]


Очікуваний результат: `[84, 102, 58, 4, 26, 88, 160, 132, 32]`

2. Використовуючи `list comprehension` cтворіть новий список `numbers_squared_lh`, де кожен елемент списку `numbers` є домноженим 2.
Так, операція, яку ми викнуємо над кожним елементом вихідного списку, та сама, що і в попередньому завданні, але інший спосіб отримання результату.

In [62]:
numbers_squared_lh = [i for i in numbers_squared]

print(numbers_squared_lh)

[84, 102, 58, 4, 26, 88, 160, 132, 32]


3. У нас вже є знайдене середнє значення за списком - воно визначене в змінній `numbers_mean`.

  Використовуючи цикл, створіть новий список `numbers_gt_mean`, який порівнює кожен елемент з цим середнім значенням за списком за наступним правилом: "якщо цей елемент спиcку `numbers` є більшим за `numbers_mean`, то запишемо 1, інакше 0". Виведіть новий список на екран.

In [63]:
numbers_mean = 38.11

In [70]:
numbers_gt_mean = []
for i in numbers_squared_lh:
    if i > numbers_mean:
      numbers_gt_mean.append(1),
    else:
      numbers_gt_mean.append(0)


In [72]:
 print(numbers_gt_mean)

[1, 1, 1, 0, 0, 1, 1, 1, 0]


Очікуваний результат: `[1, 1, 0, 0, 0, 1, 1, 1, 0]`

4. Отримайте список за тим самим правилом, що в завданні 3, тільки тепер з list comprehension. Результат запишіть у змінну `numbers_gt_mean_lc`

In [73]:
numbers_gt_mean_lc = [1 if i > numbers_mean else 0 for i in numbers_squared_lh]

In [74]:
print(numbers_gt_mean_lc)

[1, 1, 1, 0, 0, 1, 1, 1, 0]


5. Заданий список `long`. З допомогою циклу створіть словник `n_repeats`, який містить інформацію "скільки разів повторюється кожен унікальний елемент списку `long`".
Наприклад:
```
long = [1, 1, 2, 2, 3]
# тоді `n_repeats` містить {1: 2, 2: 2, 3: 1}
```

Аби для кожного унікального елементу знайти, скільки разів він повторюється, виористайте метод списку `list.count(<елемент>)`.
Виведіть `n_repeats` на екран.

**Підказки**:
1. Для початку може бути зручно отримати набір унікальних елементів списку.
2. Порядок ключів в словнику не має значення! Тобто ключі в словнику не обовʼязково мають бути в тому ж порядку, як наведено в прикладі результату. Але кількості зустрічань для кожного елемента мають співпадати з наведеними.

In [24]:
long = [75, 39, 95, 1, 91, 43, 97, 24, 86, 75, 92, 91, 15, 15, 75, 95, 54, 29, 55, 98]

In [25]:
long_set = set(long)

In [26]:
print(long_set)

{1, 97, 98, 39, 75, 43, 15, 55, 86, 54, 24, 91, 92, 29, 95}


In [78]:
n_repeats = {}

In [79]:
for element in long_set:
    n_repeats[element] = long.count(element)

In [80]:
print("Словник частоти чисел:")
print(n_repeats)

Словник частоти чисел:
{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}


Очікуваний результат:
```
{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}

```

In [1]:
words_list = [
    "Python", "контракт", "дата", "тестування", "Python", "Foundry", "код", "дані", "цикл", "словник",
    "контракт", "автоматизація", "баг", "скріншот", "Python", "дата", "аналітика", "промпт", "Midjourney", "тестування",
    "контракт", "статус", "код", "Foundry", "об'єкт", "Python", "тестування", "дата", "код", "баг",
    "змінна", "список", "контракт", "статус", "дата", "тестування", "онтологія", "фільтр", "пошук", "результат",
    "код", "цикл", "словник", "автоматизація", "скріншот", "аналітика", "промпт", "Midjourney", "статус", "Foundry",
    "об'єкт", "змінна", "список", "онтологія", "фільтр", "пошук", "результат", "дані", "QA", "Pulse",
    "звіт", "мова", "індекс", "слайсинг", "метод", "функція", "клас", "модуль", "бібліотека", "імпорт",
    "QA", "Pulse", "звіт", "мова", "індекс", "слайсинг", "метод", "функція", "клас", "модуль",
    "логіка", "умова", "рядок", "число", "істина", "хиба", "помилка", "виняток", "стек", "пам'ять",
    "логіка", "умова", "рядок", "число", "істина", "хиба", "помилка", "виняток", "стек", "пам'ять"
]

In [3]:
words_set = set(words_list)
print(words_set)

{'мова', 'словник', 'умова', 'бібліотека', 'фільтр', 'код', 'цикл', 'слайсинг', 'імпорт', 'помилка', 'дані', 'список', 'істина', 'рядок', 'автоматизація', 'стек', "пам'ять", 'онтологія', 'Pulse', 'індекс', 'Foundry', 'модуль', "об'єкт", 'промпт', 'метод', 'клас', 'баг', 'результат', 'тестування', 'звіт', 'QA', 'дата', 'пошук', 'Midjourney', 'аналітика', 'статус', 'Python', 'число', 'виняток', 'контракт', 'логіка', 'скріншот', 'хиба', 'функція', 'змінна'}


In [5]:
n_repeats = {}

In [6]:
for element in words_set:
    n_repeats[element] = words_list.count(element)

In [7]:
print("Словник частоти чисел:")
print(n_repeats)

Словник частоти чисел:
{'мова': 2, 'словник': 2, 'умова': 2, 'бібліотека': 1, 'фільтр': 2, 'код': 4, 'цикл': 2, 'слайсинг': 2, 'імпорт': 1, 'помилка': 2, 'дані': 2, 'список': 2, 'істина': 2, 'рядок': 2, 'автоматизація': 2, 'стек': 2, "пам'ять": 2, 'онтологія': 2, 'Pulse': 2, 'індекс': 2, 'Foundry': 3, 'модуль': 2, "об'єкт": 2, 'промпт': 2, 'метод': 2, 'клас': 2, 'баг': 2, 'результат': 2, 'тестування': 4, 'звіт': 2, 'QA': 2, 'дата': 4, 'пошук': 2, 'Midjourney': 2, 'аналітика': 2, 'статус': 3, 'Python': 4, 'число': 2, 'виняток': 2, 'контракт': 4, 'логіка': 2, 'скріншот': 2, 'хиба': 2, 'функція': 2, 'змінна': 2}


6. Виконайте ту саму операцію, що і в попередньому заваднні, але використовуючи `dict comprehension`. Результат запишіть у змінну `n_repeats_dh` та результат виведіть на екран.

In [11]:
n_repeats_dh = {element: words_list.count(element) for element in words_set}

In [12]:
print(n_repeats_dh)

{'мова': 2, 'словник': 2, 'умова': 2, 'бібліотека': 1, 'фільтр': 2, 'код': 4, 'цикл': 2, 'слайсинг': 2, 'імпорт': 1, 'помилка': 2, 'дані': 2, 'список': 2, 'істина': 2, 'рядок': 2, 'автоматизація': 2, 'стек': 2, "пам'ять": 2, 'онтологія': 2, 'Pulse': 2, 'індекс': 2, 'Foundry': 3, 'модуль': 2, "об'єкт": 2, 'промпт': 2, 'метод': 2, 'клас': 2, 'баг': 2, 'результат': 2, 'тестування': 4, 'звіт': 2, 'QA': 2, 'дата': 4, 'пошук': 2, 'Midjourney': 2, 'аналітика': 2, 'статус': 3, 'Python': 4, 'число': 2, 'виняток': 2, 'контракт': 4, 'логіка': 2, 'скріншот': 2, 'хиба': 2, 'функція': 2, 'змінна': 2}


In [18]:
n_repeats_dh = {element: long.count(element) for element in long_set}

In [19]:
print(n_repeats_dh)

{1: 1, 97: 1, 98: 1, 39: 1, 75: 3, 43: 1, 15: 2, 55: 1, 86: 1, 54: 1, 24: 1, 91: 2, 92: 1, 29: 1, 95: 2}


7. Знайдіть три найменші **унікальні** елементи в списку `long` та виведіть їх на екран.

In [28]:
print(long_set)

{1, 97, 98, 39, 75, 43, 15, 55, 86, 54, 24, 91, 92, 29, 95}


In [29]:
long_set_list = list(long_set)

In [31]:
long_set_list.sort()

In [34]:
result = long_set_list[:3]
print(result)

[1, 15, 24]


Очікувана відповідь: [1, 15, 24]